Fix a race that can result in unnecessary fallback executions

ivan opened the pull request:
This fixes #192.

I confirmed that the new test fails with `lib/cachex/services/courier.ex` from the previous commit, e719293bbd80c8b17b8af4e5e79102ed74ef84a7.

whitfin commented:
@ivan thanks for this. I'm still not willing to accept that this is the only way to fix it though; I'm going to look a little more into it - the server should be protecting us against race conditions. I want to rule out a bug elsewhere in the codebase.

whitfin commented on lib/cachex/services/courier.ex:
```diff
@@ -91,13 +97,7 @@ defmodule Cachex.Services.Courier do
-      Put.execute(cache, key, val, const(:notify_false))
-    end
-
+  def handle_info({ :notify, key, normalized }, { cache, tasks }) do
```
Can you please change `normalized` back to `result`?

ivan replied:
I changed what I pass in to `handle_info`; should I call `normalize_commit(result)` again there?

whitfin replied:
No, it's fine - just the name should be `result` here because there's no context as to what `normalized` means anymore :)

whitfin commented on lib/cachex/services/courier.ex:
```diff
@@ -76,7 +76,13 @@ defmodule Cachex.Services.Courier do
-            send(parent, { :notify, key, result })
+            # Commit the result immediately to avoid a race in fetch
+            # between Get.execute and Courier.dispatch.
+            normalized = normalize_commit(result)
```
For this block, please keep the same spacing from before:

```
normalized = normalize_commit(result)

with { :commit, val } <- normalized do
  Put.execute(cache, key, val, const(:notify_false))
end

send(parent, { :notify, key, normalized })
```

ivan replied:
OK, will do

whitfin commented on lib/cachex/services/courier.ex:
```diff
@@ -76,7 +76,13 @@ defmodule Cachex.Services.Courier do
                 e -> { :error, Exception.message(e) }
               end
-            send(parent, { :notify, key, result })
+            # Commit the result immediately to avoid a race in fetch
```
You can drop this comment; I'll address it later in the doc comments.

ivan replied:
OK, will do

whitfin commented on test/cachex/actions/fetch_test.exs:
```diff
@@ -95,4 +95,32 @@ defmodule Cachex.Actions.FetchTest do
+  # This test ensures that the fallback is executed just once when a
+  # fallback commit and another fetch on the same key occur simultaneously.
+  test "fetching and committing the same key simultaneously from a fallback" do
+    for duration <- 1..10 do
```
Does this set of test code really need to run 10 times? I know it's because of the change in duration, but can we not just pick `50` as a duration (or w/e), since it should be the same behaviour?

ivan replied:
Because the test doesn't always fail even when set up carefully, I run it 10 times to increase the chance of the wrong implementation failing the test. Even when run 10 times, it doesn't always fail on some machines. But the duration might not need to change, so I will investigate and maybe tweak this.

whitfin replied:
Ah, I see. Is it possible to lift the test you had in that repo you linked me to? I don't mind if the test takes a second or two.

ivan replied:
This test basically does the same thing as what was in the test repo, but is compatible with old versions of Elixir. Since you don't mind it running longer, I'll see if I can decrease its false successes by raising the number of iterations.

ivan commented:
I force-pushed with everything addressed, hopefully.  I removed the unnecessary sleeps from the test and changed `key2_fallback` to match `key1_fallback`'s execution time more closely.  Now (with the incorrect implementation) it almost always fails on the first or second iteration, but sometimes the third.  I left ten iterations in case the stars don't align repeatedly.

whitfin commented:
@ivan perfect; thank you! I'll push this up shortly.
